Migrate LiveContainerUI to SwiftUI, Support for Multiple LiveContainer, SideJIT/JITStreamer Support, Better Open in App Support, Impelement #112

hugeBlack opened the pull request:
## Switch UI to SwiftUI
I migrated most of features to an AltStore-like UI written in SwiftUI.

LiveContainerSwiftUI looks like this:
<img width="200" alt="IMG_4706" src="https://github.com/user-attachments/assets/690c2120-8b4a-4c5f-a12f-7236e47681ea">

## Support for Multiple LiveContainer
Free development accounts can install 3 apps, so now we can install 1 SideStore and 2 LiveContainers, so people can run 2 apps at the same time. App and their data, Keychains and *most* NSUserDefaults can be shared between these 2 LiveContainers. If user open a app via URL Scheme or from My Apps screen, the LiveContainer that last ran that app will be switched to.

App's data and tweaks are stored in SideStore's or AltStore's app group. If a Shared App is launched, its data will be temporarily moved to the LiveContainer's document folder to prevent crash in background because of holding a file lock in the App Group folder. It data will be moved back in last launch of LiveContainer.

Apps can't be modified when in Shared state. The second LiveContainer cannot install or manage apps.

`NSUserDefaults` are handled differently. Since it won't accept plists that moved to Library/Preferences after app launch, we have to manually read the plists and set it in `NSUserDefaults`. This solves #149 

## Better Open in App Support
It turns out that some apps don't rely on URL Schemes and use Universal Links instead. So in the new LiveContainerSwiftUI, the web page's `apple-app-site-association` will be checked and we will launch apps based on it. Then, a `NSUserActivity` will be created with these URLs and passed to `UIActivityContinuationManager`, so apps will open these URLs they support.

##  Impelement #112
Flutter apps share the bundle name of `Runner.app`, so I use bundle identifier as app bundle's name during install to avoid conflicts. Also, users can select which bundle to replace if there are multiple bundles with the same bundle identifier.

hugeBlack commented:
Users can now mark app as JIT needed and configure SideJITServer inside LiveContainer, so that apps can be launched with JIT more easily.

Also solved some weird permission issue by modifying the second one's main executable's UUID.

khanhduytran0 commented:
This is huge, at this point I may consider adding you to collaborator, thanks a lot.
While you’re at it, you may also implement #131

hugeBlack commented:
Thank you for your appreciation! I've now completed the implementation of #131. Users can now triple tap the app counter under the app list, pass the authorization and view hidden app and hide apps.

By the way, I implemented the App Clip generation, but I didn't figure out how you made `SFSafariViewController` to open data: urls and it will crash the app because it wants a http or https URL.  So I opt for starting a local server that serves that MDM config instead.

ilian reviewed (commented):
Thanks a lot for your work :muscle: This is a big PR :exploding_head: 

I didn't review the code but I wanted to give some feedback as a big fan of LiveContainer:
- I found that the `Open data folder` option when long-pressing an app is missing. I found this feature quite useful to quickly open the files app and examine the contents. It would be nice if this was kept in this PR.
- After upgrading the from d074394 (on main) to 901282bffb516bccf56be62ddeed01c7482b97e8, running an app causes its `Library/Preferences` directory to be filled with preferences of other installed apps. I tried to delete `plist` files that are not relevant to the given app but they re-appear after running the app again. Is this expected? I didn't check if this happens also for new users who did not perform an upgrade in-place.
- Each app's UUID is now less visible. I now have to hold the app and click on `Change Data Folder` to view the name of the data folder which is usually equal to the UUID if not yet renamed. It would be nice if the UUID was visible either directly under the app's name in the list (as before) or maybe show it under the app's context menu?
- The app list does not appear to have the same style as you showed in the screenshot in the description of your PR. I am running 901282b on iOS 17.6.1. I triggered a CI build for this commit by enabling GitHub Actions for my fork. You can download the IPA [under artifacts using this link](https://github.com/ilian/LiveContainer/actions/runs/10835684927). You can find a screenshot below:
![image](https://github.com/user-attachments/assets/5bd23d6c-6a46-4a4f-a357-3e4a0a80bcca)
- After installing a new app, pressing on `Run` results in the app closing and opening twice, eventually returning to the main LiveContainer app. This only seem to happen after a fresh install of the app. I am able to run the app perfectly fine after that though, even when restarting LiveContainer.

I hope that helps!

khanhduytran0 commented:
> but I didn't figure out how you made SFSafariViewController to open data: urls and it will crash the app because it wants a http or https URL.

Here’s how I bypassed it:
https://github.com/khanhduytran0/LiveContainer/blob/d074394830bd1008771f9bcf207e648814522205/LiveContainerUI/LCAppListViewController.m#L16-L21

Furthermore, generating self-signed shortcut instead is possible too, see https://github.com/0xilis/SelfSignedShortcutDemo

boa-z commented:
> Thanks a lot for your work 💪 This is a big PR 🤯
> 
> I didn't review the code but I wanted to give some feedback as a big fan of LiveContainer:
> 
> * I found that the `Open data folder` option when long-pressing an app is missing. I found this feature quite useful to quickly open the files app and examine the contents. It would be nice if this was kept in this PR.
> * After upgrading the from [d074394](https://github.com/khanhduytran0/LiveContainer/commit/d074394830bd1008771f9bcf207e648814522205) (on main) to [901282b](https://github.com/khanhduytran0/LiveContainer/commit/901282bffb516bccf56be62ddeed01c7482b97e8), running an app causes its `Library/Preferences` directory to be filled with preferences of other installed apps. I tried to delete `plist` files that are not relevant to the given app but they re-appear after running the app again. Is this expected? I didn't check if this happens also for new users who did not perform an upgrade in-place.
> * Each app's UUID is now less visible. I now have to hold the app and click on `Change Data Folder` to view the name of the data folder which is usually equal to the UUID if not yet renamed. It would be nice if the UUID was visible either directly under the app's name in the list (as before) or maybe show it under the app's context menu?
> * The app list does not appear to have the same style as you showed in the screenshot in the description of your PR. I am running [901282b](https://github.com/khanhduytran0/LiveContainer/commit/901282bffb516bccf56be62ddeed01c7482b97e8) on iOS 17.6.1. I triggered a CI build for this commit by enabling GitHub Actions for my fork. You can download the IPA [under artifacts using this link](https://github.com/ilian/LiveContainer/actions/runs/10835684927). You can find a screenshot below:
>   ![image](https://private-user-images.githubusercontent.com/25505957/367015758-5bd23d6c-6a46-4a4f-a357-3e4a0a80bcca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxODczMjAsIm5iZiI6MTcyNjE4NzAyMCwicGF0aCI6Ii8yNTUwNTk1Ny8zNjcwMTU3NTgtNWJkMjNkNmMtNmE0Ni00YTRmLWEzNTctM2U0YTBhODBiY2NhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEzVDAwMjM0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzY2UyYWMwY2NkZjQyNDk5ZDBiZjYyMmI2ZjAxM2NjM2I2M2JjZDMyNjkyMWNmMTlmMzU1NzY5ZGY2NzgwMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0WJwZ1rxMIjoRAZANyb4066U3qNpRO28WxQZiJhRObg)
> * After installing a new app, pressing on `Run` results in the app closing and opening twice, eventually returning to the main LiveContainer app. This only seem to happen after a fresh install of the app. I am able to run the app perfectly fine after that though, even when restarting LiveContainer.
> 
> I hope that helps!

I built [901282b](https://github.com/khanhduytran0/LiveContainer/commit/901282bffb516bccf56be62ddeed01c7482b97e8) locally and obtained the correct app list style. You can try the following IPA file:
[com.kdt.livecontainer_2.1.2.zip](https://github.com/user-attachments/files/16986961/com.kdt.livecontainer_2.1.2.zip)

hugeBlack commented:
Thank you for your feedback! @ilian 
- I did forget about the open data folder button. I'll add it back soon.
- It seems that the missing of Assets.car is causing the weird looking UI and missing UUID since the text color turned black and blended into the background. I presume that something went wrong when building with GitHub Actions.
- Since apps' preferences are now also stored separately and we have no way to know which plist belongs to which app, all preferences will be moved to the the preferences folder of the first app launched after the update, but should not happen after. I guess the issuse you encountered is caused by killing LiveContainer from the app switcher and remove all plists before LiveContainer's next launch. Preferences are moved back upon next launch, so I think LiveContainer just moved back all plists back that it read before you launch the app. So you may try to remove these plists after you relaunch LiveContainer and such issue should not happen again.

hugeBlack commented:
I added the "Open Data Folder" button back, change the AppClip creation method back to `data:` url, fixed GitHub Actions building issue and probably fixed the issue that apps crash after install issue. 

You can download the new version [here](https://github.com/hugeBlack/LiveContainer/actions).

> Furthermore, generating self-signed shortcut instead is possible too, see https://github.com/0xilis/SelfSignedShortcutDemo

I see. However users have to enroll a self-signed root CA to install a shortcut, which may raise some security concerns. What't more, users still have to manually select an icon before adding the generated shortcut to Home Screen, or is there a way to automate this procedure?

ilian commented:
@hugeBlack Thank you for your responses and for fixing things so quickly :rocket:

I was able to modify preferences in `Library/Preferences` and have the app read the new preferences :+1:

> all preferences will be moved to the the preferences folder of the first app launched after the update, but should not happen after

For each app, I deleted all plists that belong to other apps, click on `Run`, kill the app. Sometimes the plists came back and I had to perform this operation twice. When running the same app multiple times, I do not see any unrelated plists for that particular app. Once I start using multiple apps instead of a single one, I see unrelated plists under `Library/Preferences` again. It seems that LiveContainer is still keeping some internal state from previous runs instead of exclusively using the app's `Library/Preferences` directory as the source of truth. Are you able to reproduce this issue or do you need more detailed steps to reproduce this issue?

> probably fixed the issue that apps crash after install issue

Unfortunately, I still get the same issue after installing a new app and clicking on `Run`. After it crashes once, I am able to run it multiple times just fine.

Regarding UX, maybe it makes more sense to have the `Uninstall` item in the context menu at the bottom to be consistent with other apps such as `Files`. Maybe it also makes sense to move `Open Data Folder` to the top since I think it's more commonly used than the current item at the top (`Mark as JIT Needed`).

For context, I am running the app with version cd2c019.

hugeBlack commented:
@ilian thank for your feedback again! 👍

This commit hopefully fixed the issue that app crash once after installation (I tested on 4 devices ranging from iOS 15 to iOS 18, and they all worked fine). 

As for the mixed preference issue, after some testing, I found that the behavior of `cfprefsd` which manages `NSUserDefaults` is somewhat unpredictable and it is hard to find a stable way to dump and load preferences. So I will forgo this feature for now since I don't want to waste too much time on it and it is rather stable in spite of the mixed preference files.

What't more, I solved the issue (like #68 #123 #142) that file pickers cannot select files. The cause of this issue is that custom `UTType`s created by guest apps have not been registered in iOS so they will become nil, which made file pickers unable to select file of any type.

I intercept the call to `-[UIDocumentPickerViewController initForExportingURLs:asCopy:]` and change to `initForExportingURLs` to `@[UTTypeItem, UTTypeFolder]` so any file or folder can be chosen. Tested on Delta, UTM, iTorrent.

hugeBlack commented:
@Wei-Winter Thank for your feedback!

I think that issue with Infuse should be fixed in the latest commit. You may download and have a try. Be sure to enable 'fix file picker' setting in app settings. Do note that files will be copied to the Inbox of guest app.
